pkg/controller/controller_utils.go

```diff
@@ -1107,6 +1111,112 @@ func AddPodControllerUIDIndexer(podInformer cache.SharedIndexInformer) error {
 	})
 }
 
+// AddPodNamespaceLabelIndexer adds an indexer for Pod's namespace and labels to the given PodInformer.
+// This indexer is used to efficiently look up pods by their namespace and labels.
+func AddPodNamespaceLabelIndexer(podInformer cache.SharedIndexInformer) error {
+	if _, exists := podInformer.GetIndexer().GetIndexers()[PodNamespaceLabelIndex]; exists {
+		// indexer already exists, do nothing
+		return nil
+	}
+	return podInformer.AddIndexers(cache.Indexers{
+		PodNamespaceLabelIndex: func(obj interface{}) ([]string, error) {
+			pod, ok := obj.(*v1.Pod)
+			if !ok {
+				return nil, nil
+			}
+			// For each label in the pod, create an index key in the format of "ns:<namespace>/label:<key>=<value>"
+			indexKeys := make([]string, 0, len(pod.Labels))
+			for k, v := range pod.Labels {
+				indexKeys = append(indexKeys, fmt.Sprintf("ns:%s/label:%s=%s", pod.Namespace, k, v))
+			}
+			return indexKeys, nil
+		},
+	})
+}
+
+// GetPodsMatchingSelectorUsingPodNamespaceLabelIndex queries the index for each label selector and intersects the results efficiently.
+func GetPodsMatchingSelectorUsingPodNamespaceLabelIndex(
+	podIndexer cache.Indexer,
+	namespace string,
+	selector map[string]string,
+) ([]*v1.Pod, error) {
+	if selector == nil {
+		// Nil selector returns empty slice
+		// this should never happen other the unit test case
+		return []*v1.Pod{}, nil
+	}
+
+	if len(selector) == 0 {
+		// Empty selector matches all pods in the namespace (equivalent to labels.Everything())
+		objs, err := podIndexer.ByIndex(cache.NamespaceIndex, namespace)
+		if err != nil {
+			return nil, err
+		}
+		pods := make([]*v1.Pod, 0, len(objs))
+		for _, obj := range objs {
+			if pod, ok := obj.(*v1.Pod); ok {
```

Comment:
If type assertion fails, we should at least log something similarly like what you did in https://github.com/kubernetes/kubernetes/pull/130806/files#diff-3444629cff8d63a4b7c9f59da33ac22d6ce78b7361ae6995b84592cb24c29b45R328